### PR TITLE
Handle arbitrary amounts stdout & stderr

### PR DESCRIPTION
**Pull request by tasleson**
~Work in progress.~ Utilizing a command line wrapper around `lvm` to inject lots of additional `stderr` output, like 10MiB of text and more.

### CODE REVIEW

**Comment by tasleson** [comment]
```python
#!/usr/bin/python3

# Wrapper around lvm creating a bunch of junk stdout to stress the command
# handler in lvmdbusd
import sys
import os
import random
import string
import select
import threading
from collections import deque
from fcntl import fcntl, F_GETFL, F_SETFL
from subprocess import Popen, PIPE
import multiprocessing


run = multiprocessing.Value('i', 1)
CS = string.ascii_letters + "\n\t " + string.digits


def rs(length, character_set=CS):
    return ''.join(random.choice(character_set) for _ in range(length))


RS = rs(512)


# Make stream non-blocking
def make_non_block(stream):
    flags = fcntl(stream, F_GETFL)
    fcntl(stream, F_SETFL, flags | os.O_NONBLOCK)


def read_decoded(stream):
    tmp = stream.read()
    if tmp:
        return tmp.decode("utf-8")
    return ''


def write_random_stderr():
    amount = random.randint(1, len(RS))
    sys.stderr.write(RS[0:amount])
    sys.stderr.flush()


# Spin this off in a thread so that the main process can simply collect output
# from lvm.
def stderr_writer():
    while run.value != 0:
        write_random_stderr()


def run_cmd(args):

    writer = None
    cmd = ["/usr/sbin/lvm",]
    if args:
        cmd.extend(args)

    muck_stderr = True

    if "help" in args or "segtypes" in args:
        muck_stderr = False

    if muck_stderr:
        writer = threading.Thread(target=stderr_writer, name='write stderr')
        writer.start()

    process = Popen(cmd, stdout=PIPE, stderr=PIPE, close_fds=True,
                    env=os.environ)

    d = deque()
    stderr_text = ""
    make_non_block(process.stdout)
    make_non_block(process.stderr)

    while True:
        try:
            rd_fd = [process.stdout.fileno(), process.stderr.fileno()]
            ready = select.select(rd_fd, [], [], 2)

            for r in ready[0]:
                if r == process.stdout.fileno():
                    read = read_decoded(process.stdout)
                    d.extend(read)

                    # Write a random amount of the actual output of lvm
                    if len(d) > 0:
                        howmuch = random.randint(1, len(d))
                        for _ in range(1, howmuch):
                            c = d.popleft()
                            sys.stdout.write(c)

                        sys.stdout.flush()
                elif r == process.stderr.fileno():
                    stderr_text += read_decoded(process.stderr)

            # Check to see if process has terminated, None when running
            if process.poll() is not None:
                break
        except IOError as ioe:
            print("run_cmd:" + str(ioe))
            pass

    # Stop the write thread
    run.value = 0

    # Write everything we have left in stdout
    while len(d) > 0:
        sys.stdout.write(d.popleft())
    sys.stdout.flush()

    # Wait for thread to stop
    if muck_stderr:
        writer.join()

    # if we actually got something from lvm, dump it at the end
    sys.stderr.write(stderr_text)
    sys.stderr.flush()
    sys.exit(process.returncode)


if __name__ == "__main__":
    run_cmd(sys.argv[1:])
```

**Comment by tasleson** [comment]
@vojtechtrefny Please take a look at this PR

File: daemons/lvmdbusd/utils.py [hunk 1]
```diff
@@ -681,3 +682,16 @@ def _remove_objects(dbus_objects_rm):
 # Remove dbus objects from main thread
 def mt_remove_dbus_objects(objs):
 	MThreadRunner(_remove_objects, objs).done()
+
+
+# Make stream non-blocking
+def make_non_block(stream):
```
Comment (vojtechtrefny) [inline daemons/lvmdbusd/utils.py:688]: I was trying to figure out how to do this with `subprocess`, didn't realize calling `fnctl` on `proc.stdout` would work :man_facepalming:

**Review by vojtechtrefny — APPROVED** [review]
Looks good to me, libblockdev test suite is passing, ship it :-)

File: daemons/lvmdbusd/utils.py (hunk 1 above)
Comment (tasleson) [reply daemons/lvmdbusd/utils.py:688]: Thanks for the review and the check with `libblockdev`.  @vpodzime did a nice job coming up with the select loop in the lvmshell code.

**Comment by tasleson** [comment]
Merged upstream after I squashed it a bit: https://sourceware.org/git/?p=lvm2.git;a=commit;h=c474f174cc8b0e855f984bf211f5416b42c644a1

File: daemons/lvmdbusd/utils.py (hunk 1 above)
Comment (vpodzime) [reply daemons/lvmdbusd/utils.py:688]: Did I? :thinking:

**Pull request closed without merging**